File: shared/todo.ts
// shared/todo.ts — SINGLE SOURCE OF TRUTH for the todo wire contract (AD-6).
//
// Both `web` and `api` derive the todo shape, field names, casing, envelope, and
// success codes from this definition. `web` imports these types directly. The Go
// side (`api`) mirrors this shape and is wired to a contract test that fails the
// build on drift — completed in Story 1.2 when `POST` lands (see api/model/todo.go).
//
// Rules encoded here (do not diverge — a mismatch must be a type error, not a
// runtime bug):
//   - camelCase on the wire.
//   - timestamps are NESTED under `metadata` (never top-level).
//   - `description` is "" when empty, NEVER null.
//   - `status` is a constrained set kept in sync with the DB CHECK + Go service (AD-8).
//   - `GET /todos` returns a bare array, `[]` (never null) when empty.

/** Allowed todo lifecycle states (AD-8). Extend this list to add a state. */
export const TODO_STATUSES = ['active', 'completed'] as const;
export type TodoStatus = (typeof TODO_STATUSES)[number];

/** Server-managed timestamps, RFC3339 UTC with trailing `Z`, second precision (AD-7). */
export interface TodoMetadata {
  /** e.g. "2026-07-17T14:03:11Z" */
  createdAt: string;
  /** e.g. "2026-07-17T14:03:11Z" — updates on every create/edit/status change */
  updatedAt: string;
}

/** The canonical todo resource as it appears on the wire (AD-6). */
export interface Todo {
  /** server/DB-generated UUID v4 (AD-7) */
  id: string;
  /** required, ≤200 code points after trim (AD-10) */
  title: string;
  /** optional, ≤2000 code points after trim; "" when empty, never null (AD-10) */
  description: string;
  status: TodoStatus;
  metadata: TodoMetadata;
}

/** `POST /todos` request body. Server assigns id/status/metadata and ignores any client-supplied copies. */
export interface CreateTodoRequest {
  title: string;
  description?: string;
}

/** `PATCH /todos/:id` request body. Send ONLY changed fields; an absent field means unchanged (AD-6). */
export interface UpdateTodoRequest {
  title?: string;
  description?: string;
  status?: TodoStatus;
}

/** Uniform error envelope for every non-2xx response (AD-9). */
export interface ApiError {
  error: {
    code: 'validation_error' | 'not_found' | 'internal_error';
    message: string;
  };
}

File: backend/model/todo.go
// Package model holds the api's domain types. It is the innermost layer — every
// other layer (handler, service, repository) may depend on it, and it depends on
// nothing internal, preserving the one-way dependency spine (AD-1).
package model

// Todo mirrors the single source-of-truth wire contract in shared/todo.ts (AD-6).
// The json tags map Go's exported PascalCase to the camelCase wire shape; timestamps
// are NESTED under metadata; description is "" (never null). Any change here must be
// made in shared/todo.ts first and kept in lockstep (a contract test enforces this
// from Story 1.2).
type Todo struct {
	ID          string   `json:"id"`
	Title       string   `json:"title"`
	Description string   `json:"description"`
	Status      string   `json:"status"`
	Metadata    Metadata `json:"metadata"`
}

// Metadata carries the server-authoritative timestamps (AD-7): RFC3339 UTC with a
// trailing Z, second precision.
type Metadata struct {
	CreatedAt string `json:"createdAt"`
	UpdatedAt string `json:"updatedAt"`
}

// Allowed todo statuses (AD-8). Kept in sync with the DB CHECK constraint and, from
// Story 1.2, with service-layer validation.
const (
	StatusActive    = "active"
	StatusCompleted = "completed"
)

// APIError is the uniform non-2xx envelope (AD-9): { "error": { "code", "message" } }.
type APIError struct {
	Error APIErrorBody `json:"error"`
}

// APIErrorBody is the inner payload of APIError.
type APIErrorBody struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// Error code vocabulary (AD-9) — a fixed set the client can branch on.
const (
	CodeValidationError = "validation_error"
	CodeNotFound        = "not_found"
	CodeInternalError   = "internal_error"
)

// NewAPIError builds an APIError with the given code and message.
func NewAPIError(code, message string) APIError {
	return APIError{Error: APIErrorBody{Code: code, Message: message}}
}
